wallet/wallet-controller/src/read.rs

```diff
@@ -83,6 +83,34 @@ impl<'a, T: NodeInterface> ReadOnlyController<'a, T> {
             .map_err(ControllerError::WalletError)
     }
 
+    pub async fn get_decimal_balance(
+        &self,
+        utxo_states: UtxoStates,
+        with_locked: WithLocked,
+    ) -> Result<Balances, ControllerError<T>> {
+        let mut balances = self.get_balance(utxo_states, with_locked)?;
+
+        let coins = balances.remove(&Currency::Coin).unwrap_or(Amount::ZERO);
+        let coins = DecimalAmount::from_amount_minimal(coins, self.chain_config.coin_decimals());
+
+        let mut tokens = BTreeMap::new();
+        for (currency, amount) in balances {
+            let token_id = match currency {
+                Currency::Coin => panic!("Removed just above"),
+                Currency::Token(token_id) => token_id,
+            };
+
+            let info = super::fetch_token_info(&self.rpc_client, token_id).await?;
+            let decimals = info.token_number_of_decimals();
+            tokens.insert(
+                token_id,
+                DecimalAmount::from_amount_minimal(amount, decimals),
+            );
+        }
```

Comment:
One other option I used in the read controller for fetching multiple pool infos from the node was to use FuturesUnordered and try_collect().await to await all of them at once instead of one by one.